File: tests/php/golden/BatchResponseHandlingTest.php
<?php

declare(strict_types=1);

namespace MoySklad\Tests\Golden;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use MoySklad\Tests\TestCase;
use OpenAPI\Client\Api\ProductsApi;
use OpenAPI\Client\ApiException;
use OpenAPI\Client\Configuration;
use OpenAPI\Client\Model\BatchResponseEntity;
use OpenAPI\Client\Model\Errors;
use OpenAPI\Client\Model\Product;
use OpenAPI\Client\ObjectSerializer;

/**
 * Поведение batch-ответов:
 * 1) смешанный массив [сущность, Errors] десериализуется в BatchResponseEntity[];
 * 2) HTTP 400+ и объект Errors → ApiException;
 * 3) HTTP 400+ и массив → без ApiException, возвращается список результатов.
 */
class BatchResponseHandlingTest extends TestCase
{
    public function testMixedBatchResponseDeserializesProductAndErrors(): void
    {
        $payload = $this->loadBatchFixture('product_batch_partial.json');
        $this->assertIsArray($payload);
        $this->assertCount(2, $payload);

        $models = ObjectSerializer::deserialize(
            json_encode($payload),
            '\OpenAPI\Client\Model\BatchResponseEntity[]'
        );

        $this->assertIsArray($models);
        $this->assertCount(2, $models);
        $this->assertInstanceOf(Product::class, $models[0]);
        $this->assertInstanceOf(Errors::class, $models[1]);
        $this->assertNotEmpty($models[1]->getErrors());
        $this->assertSame(1000, $models[1]->getErrors()[0]->getCode());
    }

    public function testGlobalBatchErrorThrowsApiException(): void
    {
        $body = $this->loadBatchFixtureRaw('batch_global_error.json');
        $api = $this->productsApiWithResponse(400, $body);

        try {
            $api->createProductsBatch([$this->minimalProduct()]);
            $this->fail('Expected ApiException for global batch error object');
        } catch (ApiException $e) {
            $this->assertSame(400, $e->getCode());
            $this->assertSame($body, $e->getResponseBody());
            $this->assertStringContainsString('1056', (string) $e->getResponseBody());
        }
    }

    public function testPartialSuccessAt400ReturnsBatchWithoutException(): void
    {
        $body = $this->loadBatchFixtureRaw('product_batch_partial.json');
        $api = $this->productsApiWithResponse(400, $body);

        $result = $api->createProductsBatch([$this->minimalProduct()]);

        $this->assertIsArray($result);
        $this->assertCount(2, $result);
        $this->assertContainsOnlyInstancesOf(BatchResponseEntity::class, $result);
        $this->assertInstanceOf(Product::class, $result[0]);
        $this->assertInstanceOf(Errors::class, $result[1]);
        $this->assertSame(1000, $result[1]->getErrors()[0]->getCode());
    }

    private function productsApiWithResponse(int $status, string $body): ProductsApi
    {
        $mock = new MockHandler([
            new Response($status, ['Content-Type' => 'application/json'], $body),
        ]);
        $client = new Client(['handler' => HandlerStack::create($mock)]);
        $config = Configuration::getDefaultConfiguration()
            ->setHost('http://batch-test.local/api/remap/1.2');

        return new ProductsApi($client, $config);
    }

    private function minimalProduct(): Product
    {
        $product = new Product();
        $product->setName('Batch test product');

        return $product;
    }

    /** @return array<mixed> */
    private function loadBatchFixture(string $filename): array
    {
        $path = $this->getBatchFixturesPath() . '/' . $filename;
        $this->assertFileExists($path, "Batch fixture not found: {$filename}");

        $data = json_decode((string) file_get_contents($path), true);
        $this->assertSame(JSON_ERROR_NONE, json_last_error(), json_last_error_msg());

        return $data;
    }

    private function loadBatchFixtureRaw(string $filename): string
    {
        $path = $this->getBatchFixturesPath() . '/' . $filename;
        $this->assertFileExists($path, "Batch fixture not found: {$filename}");

        return (string) file_get_contents($path);
    }

    private function getBatchFixturesPath(): string
    {
        return $this->getFixturesPath() . '/batch';
    }
}

File: tests/java/assertions/src/test/java/com/lognex/test/golden/BatchResponseHandlingTest.java
package com.lognex.test.golden;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.lognex.test.BaseTestCase;
import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import ru.moysklad.remap_1_2.ApiClient;
import ru.moysklad.remap_1_2.ApiException;
import ru.moysklad.remap_1_2.model.BatchResponseEntity;
import ru.moysklad.remap_1_2.model.Errors;
import ru.moysklad.remap_1_2.model.Product;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;

/**
 * Поведение batch-ответов:
 * 1) смешанный массив [сущность, Errors] десериализуется в List&lt;BatchResponseEntity&gt;;
 * 2) HTTP 400+ и объект Errors → ApiException;
 * 3) HTTP 400+ и массив → без ApiException, возвращается список результатов.
 *
 * HTTP-сценарии вызывают {@link ApiClient#invokeAPIBatch} напрямую
 * (тот же путь, что у batch API-методов), чтобы не упираться в известный
 * Hibernate Validator конфликт иерархии Product/ProductMarker.
 */
class BatchResponseHandlingTest extends BaseTestCase {

    private static final ObjectMapper MAPPER = new ApiClient().getObjectMapper();
    private static final String BATCH_PATH = "/entity/product/batch";

    @Test
    void testMixedBatchResponseDeserializesProductAndErrors() throws Exception {
        String json = loadBatchFixtureRaw("product_batch_partial.json");
        List<BatchResponseEntity> models = MAPPER.readValue(
                json,
                new TypeReference<List<BatchResponseEntity>>() {}
        );

        Assertions.assertEquals(2, models.size());
        Assertions.assertInstanceOf(Product.class, models.get(0));
        Assertions.assertInstanceOf(Errors.class, models.get(1));
        Errors errors = (Errors) models.get(1);
        Assertions.assertFalse(errors.getErrors().isEmpty());
        Assertions.assertEquals(1000, errors.getErrors().get(0).getCode());
    }

    @Test
    void testGlobalBatchErrorThrowsApiException() throws Exception {
        String body = loadBatchFixtureRaw("batch_global_error.json");

        try (LocalBatchServer server = LocalBatchServer.start(400, body)) {
            ApiClient client = apiClient(server.baseUrl());
            ApiException exception = Assertions.assertThrows(
                    ApiException.class,
                    () -> invokeProductsBatch(client)
            );
            Assertions.assertEquals(400, exception.getCode());
            Assertions.assertTrue(
                    exception.getResponseBody().contains("\"code\":1056")
                            || exception.getResponseBody().contains("\"code\": 1056"),
                    "response body should contain error code 1056: " + exception.getResponseBody()
            );
            Assertions.assertNotNull(exception.getApiErrors());
            Assertions.assertFalse(exception.getApiErrors().getErrors().isEmpty());
            Assertions.assertEquals(1056, exception.getApiErrors().getErrors().get(0).getCode());
        }
    }

    @Test
    void testPartialSuccessAt400ReturnsBatchWithoutException() throws Exception {
        String body = loadBatchFixtureRaw("product_batch_partial.json");

        try (LocalBatchServer server = LocalBatchServer.start(400, body)) {
            List<BatchResponseEntity> result = invokeProductsBatch(apiClient(server.baseUrl()));

            Assertions.assertEquals(2, result.size());
            Assertions.assertInstanceOf(Product.class, result.get(0));
            Assertions.assertInstanceOf(Errors.class, result.get(1));
            Errors errors = (Errors) result.get(1);
            Assertions.assertEquals(1000, errors.getErrors().get(0).getCode());
        }
    }

    private ApiClient apiClient(String baseUrl) {
        ApiClient client = new ApiClient();
        client.setBasePath(baseUrl);
        return client;
    }

    private List<BatchResponseEntity> invokeProductsBatch(ApiClient client) throws ApiException {
        Map<String, Object> product = new HashMap<String, Object>();
        product.put("name", "Batch test product");

        return client.invokeAPIBatch(
                BATCH_PATH,
                "POST",
                Collections.emptyList(),
                Collections.emptyList(),
                null,
                Collections.singletonList(product),
                new HashMap<String, String>(),
                new HashMap<String, String>(),
                new HashMap<String, Object>(),
                "application/json",
                "application/json",
                new String[0],
                new TypeReference<List<BatchResponseEntity>>() {}
        );
    }

    private String loadBatchFixtureRaw(String filename) throws URISyntaxException, IOException {
        Path path = getFixturesPath().resolve("batch").resolve(filename);
        Assertions.assertTrue(Files.exists(path), "Batch fixture not found: " + filename);
        return new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
    }

    private static final class LocalBatchServer implements AutoCloseable {
        private final HttpServer server;
        private final String baseUrl;

        private LocalBatchServer(HttpServer server, String baseUrl) {
            this.server = server;
            this.baseUrl = baseUrl;
        }

        static LocalBatchServer start(int status, String body) throws IOException {
            HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
            byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
            server.createContext(BATCH_PATH, exchange -> {
                try {
                    // Drain request body so the client can finish sending (Java 8 compatible).
                    drain(exchange.getRequestBody());
                    exchange.getResponseHeaders().add("Content-Type", "application/json");
                    exchange.sendResponseHeaders(status, bytes.length);
                    try (OutputStream os = exchange.getResponseBody()) {
                        os.write(bytes);
                    }
                } finally {
                    exchange.close();
                }
            });
            server.setExecutor(Executors.newSingleThreadExecutor());
            server.start();
            String baseUrl = "http://127.0.0.1:" + server.getAddress().getPort();
            return new LocalBatchServer(server, baseUrl);
        }

        String baseUrl() {
            return baseUrl;
        }

        @Override
        public void close() {
            server.stop(0);
        }

        private static void drain(InputStream in) throws IOException {
            byte[] buffer = new byte[1024];
            while (in.read(buffer) != -1) {
                // discard
            }
        }
    }
}
